gym-crowdedness.html

```diff
@@ -0,0 +1,383 @@
+
+%% md
+We want to explore the correlations between the different columns. To do so, we will need to calculate the correlation coefficients with the following function, which accepts two feature columns as parameters and returns their Pearson Correlation Coefficient in the range [-1, 1]
+
+%% js
+/*
+ *  Source: http://stevegardner.net/2012/06/11/javascript-code-to-calculate-the-pearson-correlation-coefficient/
+ */
+function getPearsonCorrelation(x, y) {
+    var shortestArrayLength = 0;
+     
+    if(x.length == y.length) {
+        shortestArrayLength = x.length;
+    } else if(x.length > y.length) {
+        shortestArrayLength = y.length;
+        console.error('x has more items in it, the last ' + (x.length - shortestArrayLength) + ' item(s) will be ignored');
+    } else {
+        shortestArrayLength = x.length;
+        console.error('y has more items in it, the last ' + (y.length - shortestArrayLength) + ' item(s) will be ignored');
+    }
+  
+    var xy = [];
+    var x2 = [];
+    var y2 = [];
+  
+    for(var i=0; i<shortestArrayLength; i++) {
+        xy.push(x[i] * y[i]);
+        x2.push(x[i] * x[i]);
+        y2.push(y[i] * y[i]);
+    }
+  
+    var sum_x = 0;
+    var sum_y = 0;
+    var sum_xy = 0;
+    var sum_x2 = 0;
+    var sum_y2 = 0;
+  
+    for(var i=0; i< shortestArrayLength; i++) {
+        sum_x += x[i];
+        sum_y += y[i];
+        sum_xy += xy[i];
+        sum_x2 += x2[i];
+        sum_y2 += y2[i];
+    }
+  
+    var step1 = (shortestArrayLength * sum_xy) - (sum_x * sum_y);
+    var step2 = (shortestArrayLength * sum_x2) - (sum_x * sum_x);
+    var step3 = (shortestArrayLength * sum_y2) - (sum_y * sum_y);
+    var step4 = Math.sqrt(step2 * step3);
+    var answer = step1 / step4;
+  
+    return answer;
+}
+
+%% md
+Notice that the 'date' and 'timestamp' feature columns are continuous and will not give us meaningful correlations. Let's delete them.
+
+%% js
+delete columns['date']
+delete columns['timestamp']
+
+%% md
+We are now ready to calculate the correlation coefficients for all column pairs
+
+%% js
+var correlationCoefficients = {}
+
+Object.keys(columns).forEach(function(xColName) {
+	var xColumn = columns[xColName]
+    correlationCoefficients[xColName] = new Array()
+    Object.keys(columns).forEach(function(yColName) {
+      var yColumn = columns[yColName]
+      correlationCoefficients[xColName].push(getPearsonCorrelation(xColumn, yColumn))
+    })
+})
+
+%% md
+## Visualize correlations using plotly.js heatmap
+
+%% md
+<div id='plot'></div>
+
+%% resource
+https://cdn.plot.ly/plotly-latest.min.js
+
+%% md
+Plot the correlations in a heatmap
+
+%% js
+var zValues = new Array()
+Object.keys(correlationCoefficients).forEach(function(ccKey) {
+  zValues.push(correlationCoefficients[ccKey])
+})
+
+var heatmapData = [
+  {
+    z: zValues,
+    x: Object.keys(correlationCoefficients),
+    y: Object.keys(correlationCoefficients),
+    colorscale: 'Greens',
```

Comment:
Is there a colorscale where the negative values are red and the positive values are blue? I can't tell what is positively or negatively correlated in this heatmap. Only upon seeing your conclusions did I understand that I had been misreading the chart on first glance.

Reply:
+1. You probably want to use a "divergent" color scale that has 0 as the neutral point. see:
https://plot.ly/javascript/colorscales/
'Bluered' or 'RdBu' are obvious choices, but 'portland' or 'picnic' would be ok too if you think they end up looking better.